File: horizons_py/horizons/actions.py
from __future__ import annotations

from typing import Any, Dict, List, Optional
from uuid import UUID

from .client import HorizonsClient
from . import models


class ActionsAPI:
    def __init__(self, client: HorizonsClient) -> None:
        self._client = client

    async def propose(
        self,
        *,
        agent_id: str,
        action_type: str,
        payload: Any,
        risk_level: models.RiskLevel,
        dedupe_key: Optional[str] = None,
        context: Any,
        ttl_seconds: Optional[int] = None,
        project_id: Optional[UUID] = None,
    ) -> UUID:
        body: Dict[str, Any] = {
            "agent_id": agent_id,
            "action_type": action_type,
            "payload": payload,
            "risk_level": risk_level.value,
            "context": context,
        }
        if dedupe_key is not None:
            body["dedupe_key"] = dedupe_key
        if ttl_seconds is not None:
            body["ttl_seconds"] = ttl_seconds
        if project_id:
            body["project_id"] = str(project_id)
        resp = await self._client._request("POST", "/api/v1/actions/propose", json=body)
        data = await self._client.json(resp)
        return UUID(data["action_id"])

    async def approve(
        self, action_id: UUID, *, reason: str, project_id: Optional[UUID] = None
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {"reason": reason}
        if project_id:
            body["project_id"] = str(project_id)
        resp = await self._client._request(
            "POST", f"/api/v1/actions/{action_id}/approve", json=body
        )
        return await self._client.json(resp)

    async def deny(
        self, action_id: UUID, *, reason: str, project_id: Optional[UUID] = None
    ) -> Dict[str, Any]:
        body: Dict[str, Any] = {"reason": reason}
        if project_id:
            body["project_id"] = str(project_id)
        resp = await self._client._request(
            "POST", f"/api/v1/actions/{action_id}/deny", json=body
        )
        return await self._client.json(resp)

    async def pending(
        self, *, project_id: Optional[UUID] = None, limit: int = 100, offset: int = 0
    ) -> List[models.ActionProposal]:
        params: Dict[str, Any] = {"limit": limit, "offset": offset}
        if project_id:
            params["project_id"] = str(project_id)
        resp = await self._client._request("GET", "/api/v1/actions/pending", params=params)
        data = await self._client.json(resp)
        return [models.ActionProposal.model_validate(item) for item in data]

File: horizons_ts/src/actions.ts
import { HorizonsClient } from "./client";
import { ActionProposal, RiskLevel, UUID } from "./types";

/** Action lifecycle API (propose, approve, deny, list pending). */
export class ActionsAPI {
  constructor(private client: HorizonsClient) {}

  /** Propose an action for approval and return the action ID. */
  async propose(input: {
    agent_id: string;
    action_type: string;
    payload: unknown;
    risk_level: RiskLevel;
    dedupe_key?: string;
    context: unknown;
    ttl_seconds?: number;
    project_id?: string;
  }): Promise<UUID> {
    const resp = await this.client.request<{ action_id: UUID }>("/api/v1/actions/propose", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(input),
    });
    return resp.action_id;
  }

  /** Approve a pending action proposal. */
  async approve(actionId: UUID, reason: string, project_id?: string): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = { reason };
    if (project_id) body.project_id = project_id;
    return await this.client.request(`/api/v1/actions/${actionId}/approve`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(body),
    });
  }

  /** Deny a pending action proposal. */
  async deny(actionId: UUID, reason: string, project_id?: string): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = { reason };
    if (project_id) body.project_id = project_id;
    return await this.client.request(`/api/v1/actions/${actionId}/deny`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(body),
    });
  }

  /** List pending action proposals for the project/org scope. */
  async pending(limit = 100, offset = 0, project_id?: string): Promise<ActionProposal[]> {
    const params = new URLSearchParams({ limit: String(limit), offset: String(offset) });
    if (project_id) params.append("project_id", project_id);
    const path = `/api/v1/actions/pending?${params.toString()}`;
    return await this.client.request<ActionProposal[]>(path, { method: "GET" });
  }
}
